lib/internal/dns/promises.js

```diff
@@ -0,0 +1,250 @@
+'use strict';
+const {
+  bindDefaultResolver,
+  getDefaultResolver,
+  Resolver: Resolver_,
+  validateHints
+} = require('internal/dns/utils');
+const { codes, dnsException } = require('internal/errors');
+const { isIP, isIPv4, isLegalPort } = require('internal/net');
+const {
+  getaddrinfo,
+  getnameinfo,
+  ChannelWrap,
+  GetAddrInfoReqWrap,
+  GetNameInfoReqWrap,
+  QueryReqWrap
+} = process.binding('cares_wrap');
+const {
+  ERR_INVALID_ARG_TYPE,
+  ERR_INVALID_OPT_VALUE,
+  ERR_MISSING_ARGS,
+  ERR_SOCKET_BAD_PORT
+} = codes;
+
+
+function onlookup(err, addresses) {
+  if (err) {
+    this.reject(dnsException(err, 'getaddrinfo', this.hostname));
+    return;
+  }
+
+  const family = this.family ? this.family : isIPv4(addresses[0]) ? 4 : 6;
+  this.resolve({ address: addresses[0], family });
+}
+
+function onlookupall(err, addresses) {
+  if (err) {
+    this.reject(dnsException(err, 'getaddrinfo', this.hostname));
+    return;
+  }
+
+  const family = this.family;
+
+  for (var i = 0; i < addresses.length; i++) {
+    const address = addresses[i];
+
+    addresses[i] = {
+      address,
+      family: family ? family : isIPv4(addresses[i]) ? 4 : 6
+    };
+  }
+
+  this.resolve(addresses);
+}
+
+function createLookupPromise(family, hostname, all, hints, verbatim) {
+  return new Promise((resolve, reject) => {
+    if (!hostname) {
+      if (all)
+        resolve([]);
+      else
+        resolve({ address: null, family: family === 6 ? 6 : 4 });
+
+      return;
+    }
+
+    const matchedFamily = isIP(hostname);
+
+    if (matchedFamily !== 0) {
+      const result = { address: hostname, family: matchedFamily };
+      if (all)
+        resolve([result]);
+      else
+        resolve(result);
+
+      return;
+    }
+
+    const req = new GetAddrInfoReqWrap();
+
+    req.family = family;
+    req.hostname = hostname;
+    req.oncomplete = all ? onlookupall : onlookup;
+    req.resolve = resolve;
+    req.reject = reject;
```

Comment:
Right now we seem to have two implicit "subclasses" of `GetNameInfoReqWrap`: one having a `.callback` property and is used for `dns`, while the other has `.resolve` and `.reject` properties. I would rather make this distinction explicit by having a `GetAddrInfoPromiseReqWrap` subclass that extends `GetAddrInfoReqWrap`.

Same for the other wrapper classes.

Reply:
@TimothyGu I was aiming to avoid a lot of unnecessary duplicate code. The wraps behave exactly the same in the C++ layer.

Reply:
@cjihrig Yeah I understand. What I'm trying to say is that you can have a JavaScript `class` that `extends GetAddrInfoReqWrap`. As long as its constructor calls `super()` properly it should work just fine.

Reply:
@TimothyGu It still seems a bit complex if we hang all the data (`host`, `port`, `bindingName`, etc.) onto a req wrap with C++ binding even though the C++ binding may not necessarily read those properties from the req wrap. The only property matters to C++ should be `. oncomplete`, so it may be cleaner if we create two JS classes, one for the callback and the other one for promises, each with a `.handle` (or a symbol property) pointing to those req wraps and add the data properties to the JS instances of those classes. Then the main difference between those two classes would just be how we call `resolve`/`reject`/`callback(null, result)`/`callback(dnsException(...))` in `handle.oncomplete`.

(It's a bit unfortunate that the resolves have a `ChannelWrap` that's not req wrap as `_handle` though)

Reply:
On a second thought, it may not be feasible to separate the data properties to an outer object since we have been allowing users to retrieve them via the `this` bound to the callback...